### PR TITLE
[FLINK-15137][avro] Improve schema derivation for Avro format

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Support schema derivation for Avro format and also enables schema derivation as the default behavior for Avro format, so that users doesn't need to specify a format schema explicitly.

## Brief change log

The commit messages describe the changes.

## Verifying this change

- add tests to verify converting a `LogicalType` to `Schema` in `AvroSchemaConverterTest`.
- add tests to verify the `(De)SerializationSchema` created from `LogicalType` works well in `AvroRowDeSerializationSchemaTest`.
- add tests to verify that the derived schema can work both on serialization and deserialization in `AvroRowFormatFactoryTest`. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
Hi @twalthr , I rebased this PR as #10500 is merged. I'm appreciate if you can review this when you are free.

**Comment by twalthr** [comment]
Sure, I will take a look.

**Comment by wuchong** [comment]
@flinkbot run travis

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 1]
```diff
@@ -218,67 +249,65 @@ private Object convertAvroType(Schema schema, TypeInformation<?> info, Object ob
...
-				if (info instanceof BasicArrayTypeInfo) {
-					final TypeInformation<?> elementInfo = ((BasicArrayTypeInfo<?, ?>) info).getComponentInfo();
-					return convertToObjectArray(schema.getElementType(), elementInfo, object);
-				} else {
-					final TypeInformation<?> elementInfo = ((ObjectArrayTypeInfo<?, ?>) info).getComponentInfo();
-					return convertToObjectArray(schema.getElementType(), elementInfo, object);
-				}
+				LogicalType elementType = ((ArrayType) type).getElementType();
+				// transform to sql time conversion classes
+				DataType elementDataType = DataTypeUtils.transform(
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: Here were are in the hot path of the runtime code. We should avoid creating objects. IMHO this is another example that API classes (everything outside of planner) should use `DataType` instead. This adds no runtime overhead but only a bit of pre-flight overhead.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 2]
```diff
@@ -196,19 +225,21 @@ public int hashCode() {
...
 			final Schema.Field field = fields.get(i);
-			row.setField(i, convertAvroType(field.schema(), fieldInfo[i], record.get(i)));
+			row.setField(i, convertAvroType(field.schema(), fieldTypes.get(i), record.get(i)));
 		}
 		return row;
 	}
 
-	private Object convertAvroType(Schema schema, TypeInformation<?> info, Object object) {
+	private Object convertAvroType(Schema schema, LogicalType type, Object object) {
 		// we perform the conversion based on schema information but enriched with pre-computed
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: update comment

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 3]
```diff
@@ -196,19 +225,21 @@ public int hashCode() {
 
 	// --------------------------------------------------------------------------------------------
 
-	private Row convertAvroRecordToRow(Schema schema, RowTypeInfo typeInfo, IndexedRecord record) {
+	private Row convertAvroRecordToRow(Schema schema, RowType rowType, IndexedRecord record) {
 		final List<Schema.Field> fields = schema.getFields();
-		final TypeInformation<?>[] fieldInfo = typeInfo.getFieldTypes();
+		final List<LogicalType> fieldTypes = rowType.getFields().stream()
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: use `getChildren()`?

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java [hunk 4]
```diff
@@ -157,4 +209,162 @@ private AvroSchemaConverter() {
...
+						field.name(),
+						convertToDataType(field.schema()));
+				}
+				return DataTypes.ROW(dataTypeFields);
+			case ENUM:
+			case STRING:
+				// convert Avro's Utf8/CharSequence to String
+				return DataTypes.STRING();
+			case ARRAY:
+				// result type might either be ObjectArrayTypeInfo or BasicArrayTypeInfo for Strings
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: remove comment

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java [hunk 5]
```diff
@@ -157,4 +209,162 @@ private AvroSchemaConverter() {
...
+				return DataTypes.ARRAY(convertToDataType(schema.getElementType()));
+			case MAP:
+				return DataTypes.MAP(DataTypes.STRING(), convertToDataType(schema.getValueType()));
+			case UNION:
+				final Schema actualSchema;
+				if (schema.getTypes().size() == 2 && schema.getTypes().get(0).getType() == Schema.Type.NULL) {
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: How do we deal with nullability in Avro? We have a type system that supports nullability now. And Avro also has this concept. We should move this part to a separate method and call `notNull()`,`nullable()` on the resulting data type.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java [hunk 6]
```diff
@@ -157,4 +209,162 @@ private AvroSchemaConverter() {
...
+				return SchemaBuilder.builder().nullable().bytesType();
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
+				// use long to represents Timestamp
+				return LogicalTypes.timestampMillis().addToSchema(SchemaBuilder.builder().longType());
+			case DATE:
+				// use int to represents Date
+				return LogicalTypes.date().addToSchema(SchemaBuilder.builder().intType());
+			case TIME_WITHOUT_TIME_ZONE:
+				// use int to represents Time, we only support millisecond when deserialization
+				return LogicalTypes.timeMillis().addToSchema(SchemaBuilder.builder().intType());
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: Can't we support more than millis now?

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java [hunk 7]
```diff
@@ -157,4 +209,162 @@ private AvroSchemaConverter() {
...
+				final org.apache.avro.LogicalType logicalType = schema.getLogicalType();
+				if (logicalType == LogicalTypes.date()) {
+					return DataTypes.DATE();
+				} else if (logicalType == LogicalTypes.timeMillis()) {
+					return DataTypes.TIME(3);
+				}
+				return DataTypes.INT();
+			case LONG:
+				// logical timestamp type
+				if (schema.getLogicalType() == LogicalTypes.timestampMillis()) {
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: Add also micros?

**Pull request closed without merging**